### PR TITLE
[ASCII-998] Create PR on "Update buildimages" github workflow invoke

### PR DESCRIPTION
**Pull request by misteriaud**
### What does this PR do?

This PR aims to creates a PR when ["Update buildimages" github workflow](https://github.com/DataDog/datadog-agent/actions/workflows/buildimages-update.yml) is invoked.

### Motivation

This PR takes place in the automatisation of the Golang update process.

### CODE REVIEW

File: .github/workflows/buildimages-update.yml [hunk 1]
```diff
@@ -73,6 +72,11 @@ jobs:
         id: current_go_version
         run: |
           echo "GO_VERSION=$(inv go-version)" >> $GITHUB_OUTPUT
+    
+      - name: Get current buildimage tag
+        id: current_buildimage_tag
+        run: |
+            echo "BUILDIMAGE_TAG=$(inv pipeline.get-image-tag)" >> $GITHUB_OUTPUT
```
Comment (GustavoCaso) [inline .github/workflows/buildimages-update.yml:79]: Do we need to run this step to store the image_tag only to later be use in one place?

Could we inline this in the  `Create PR` step?

Comment (pgimalac) [reply .github/workflows/buildimages-update.yml:79]: This step fetches the current buildimage Id, and the next step updates it, we have to query it before it's updated

File: .github/workflows/buildimages-update.yml [hunk 2]
```diff
@@ -100,3 +104,35 @@ jobs:
...
           # if they exist, so we can skip the fetch
           skip_fetch: true
+
+      - name: Check if PR exists
+        id: check_pr
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          base_branch=${{ github.ref }} # Get the full ref
+          base_branch=${base_branch#refs/heads/} # Remove 'refs/heads/' from the ref
```
Comment (GustavoCaso) [inline .github/workflows/buildimages-update.yml:114]: NIt 🪄

File: tasks/buildimages.py [hunk 3]
```diff
@@ -33,3 +33,39 @@ def update_test_infra_definitions(ctx: Context, commit_sha: str, go_mod_only: bo
...
+    Generate the PR body used for buildimages-update Github workflow
+    """
+    pr_body = f"""This PR was automatically created by the [Update buildimages Github Workflow](https://github.com/DataDog/datadog-agent/actions/workflows/buildimages-update.yml).
+
+### Buildimages update
+This PR updates the current buildimages (`{old_build_image_tag}`) to `{new_build_image_tag}`{'(test version)' if test_version else ''}, here is the full changelog between:
+https://github.com/DataDog/datadog-agent-buildimages/compare/{old_build_image_tag.split("-")[1]}...{new_build_image_tag.split("-")[1]}
+"""
+
+    if old_go_version != new_go_version:
```
Comment (GustavoCaso) [inline tasks/buildimages.py:65]: Would there be cases that both version are the same?

File: .github/workflows/buildimages-update.yml (hunk 1 above)
Comment (GustavoCaso) [reply .github/workflows/buildimages-update.yml:79]: Thank for the explanation

File: tasks/buildimages.py (hunk 3 above)
Comment (misteriaud) [reply tasks/buildimages.py:65]: If the purpose of the new buildimage is not a Golang update, Go version remain the same (eg #26730).

File: .github/workflows/buildimages-update.yml
[the remark this answers is not in the document]
Comment (misteriaud) [reply .github/workflows/buildimages-update.yml:1]: Checking if the PR body hasn't been changed is not so easy and might adding some complexity for minimal gain IMO.
I think letting the user update the PR stay a good solution.

**Review by f4usto — COMMENTED** [review]
@maxime, just one question as I am not very familiar with this workflow.
Will this be manually triggered?

File: .github/workflows/buildimages-update.yml (hunk 1 above)
Comment (KevinFairise2) [reply .github/workflows/buildimages-update.yml:79]: I think you should not rely on the version defined in circle CI config file. We plan to get rid of circle CI and it would probably better to fetch this version in `.gitlab-ci.yml`

File: .github/workflows/buildimages-update.yml [hunk 4]
```diff
@@ -100,3 +104,43 @@ jobs:
...
+          ${{ steps.current_go_version.outputs.GO_VERSION }} \
+          ${{ inputs.go_version }} \
+          ${{ inputs.test_version && '--test-version' || '' }} > $TMP_PR_BODY_PATH
+
+          gh pr create \ # Create the PR
+          -B ${{ github.ref }} \
+          -t $PR_TITLE \
+          -F $TMP_PR_BODY_PATH \
+          -l $PR_LABELS \
+          -d \
```
Comment (KevinFairise2) [inline .github/workflows/buildimages-update.yml:146]:
```suggestion
          -d \

```

File: tasks/buildimages.py [hunk 5]
```diff
@@ -33,3 +33,39 @@ def update_test_infra_definitions(ctx: Context, commit_sha: str, go_mod_only: bo
...
+):
+    """
+    Generate the PR body used for buildimages-update Github workflow
+    """
+    pr_body = f"""This PR was automatically created by the [Update buildimages Github Workflow](https://github.com/DataDog/datadog-agent/actions/workflows/buildimages-update.yml).
+
+### Buildimages update
+This PR updates the current buildimages (`{old_build_image_tag}`) to `{new_build_image_tag}`{'(test version)' if test_version else ''}, here is the full changelog between:
+https://github.com/DataDog/datadog-agent-buildimages/compare/{old_build_image_tag.split("-")[1]}...{new_build_image_tag.split("-")[1]}
+"""
```
Comment (amenasria) [inline tasks/buildimages.py:63]: **💬 suggestion:** ‏This is very very hard to read imo. There are 2 things that are confusing my brain:
- The ``to `{new_build_image_tag}`{'(test version)'`` especially the ``}`{'(`` destroyed my brain. To fix this I would put the ``{'(test version)' if test_version else ''}`` in its own variable
- The last inline url would be fine if it wasn't a multiline string but as is it's hard to read. To fix this I would put the whole url in a variable outside of the link.

TL;DR my suggestion:

```suggestion
    buildimages_workflow_url = "https://github.com/DataDog/datadog-agent/actions/workflows/buildimages-update.yml"
    test_version_str = "(test version)" if test_version else ""
    compare_url = f"https://github.com/DataDog/datadog-agent-buildimages/compare/{old_build_image_tag.split('-')[1]}...{new_build_image_tag.split('-')[1]}"
    pr_body = f"""This PR was automatically created by the [Update buildimages Github Workflow]({buildimages_workflow_url}).

### Buildimages update
This PR updates the current buildimages (`{old_build_image_tag}`) to `{new_build_image_tag}`{test_version_str}, [here is the full changelog]({compare_url}).
"""
```

File: tasks/buildimages.py [hunk 6]
```diff
@@ -33,3 +33,39 @@ def update_test_infra_definitions(ctx: Context, commit_sha: str, go_mod_only: bo
...
+
+    if old_go_version != new_go_version:
+        pr_body += f"""
+
+### Golang update
+This PR updates the current Golang version ([`{old_go_version}`](https://go.dev/doc/devel/release#go{old_go_version})) to [`{new_go_version}`](https://go.dev/doc/devel/release#go{new_go_version}).
+"""
+    print(pr_body)
```
Comment (amenasria) [inline tasks/buildimages.py:71]: **💬 suggestion:** ‏It might be better to return the value instead and to set the `autoprint=True` in the `@task` decorator to automatically print the return value when called through the cli. WDYT?
```suggestion
    return pr_body
```

File: tasks/buildimages.py [hunk 7]
```diff
@@ -33,3 +33,39 @@ def update_test_infra_definitions(ctx: Context, commit_sha: str, go_mod_only: bo
...
+
+@task(
+    help={
+        "old_build_image_tag": "The old build image tag",
+        "new_build_image_tag": "The new build image tag",
+        "old_go_version": "The old Go version",
+        "new_go_version": "The new Go version",
+        "test_version": "Flag to indicate if this is a test version",
+    }
+)
```
Comment (amenasria) [inline tasks/buildimages.py:46]:
```suggestion
@task(
    help={
        "old_build_image_tag": "The old build image tag",
        "new_build_image_tag": "The new build image tag",
        "old_go_version": "The old Go version",
        "new_go_version": "The new Go version",
        "test_version": "Flag to indicate if this is a test version",
    },
    autoprint=True
)
```

See the comment below about the `return`

File: tasks/pipeline.py [hunk 8]
```diff
@@ -810,6 +810,24 @@ def update_circleci_config(file_path, image_tag, test_version):
         circle.write(circle_ci.replace(f"{match.group(0)}", f"{image}:{image_tag}\n"))
 
 
+@task(
+    help={
+        "file_path": "path of the CircleCI configuration YAML file",
+    }
+)
```
Comment (amenasria) [inline tasks/pipeline.py:817]:
```suggestion
@task(
    help={
        "file_path": "path of the CircleCI configuration YAML file",
    },
    autoprint=True
)
```


**💬 suggestion:** ‏Same as above

File: tasks/pipeline.py [hunk 9]
```diff
@@ -810,6 +810,24 @@ def update_circleci_config(file_path, image_tag, test_version):
...
+    """
+    Print the current image tag of the given CircleCI configuration file (default: ".circleci/config.yml")
+    """
+    image_name = "gcr.io/datadoghq/agent-circleci-runner"
+    with open(file_path) as circle:
+        circle_ci = circle.read()
+    match = re.search(rf"({image_name}(_test_only)?):([a-zA-Z0-9_-]+)\n", circle_ci)
+    if not match:
+        raise RuntimeError(f"Impossible to find the version of image {image_name} in circleci configuration file")
+    print(match.group(0).split(":")[1].strip())
```
Comment (amenasria) [inline tasks/pipeline.py:828]:
```suggestion
    return match.group(0).split(":")[1].strip()
```


**💬 suggestion:** ‏Same as above

File: tasks/pipeline.py [hunk 10]
```diff
@@ -810,6 +810,24 @@ def update_circleci_config(file_path, image_tag, test_version):
...
+def get_image_tag(_, file_path=".circleci/config.yml"):
+    """
+    Print the current image tag of the given CircleCI configuration file (default: ".circleci/config.yml")
+    """
+    image_name = "gcr.io/datadoghq/agent-circleci-runner"
+    with open(file_path) as circle:
+        circle_ci = circle.read()
+    match = re.search(rf"({image_name}(_test_only)?):([a-zA-Z0-9_-]+)\n", circle_ci)
+    if not match:
+        raise RuntimeError(f"Impossible to find the version of image {image_name} in circleci configuration file")
```
Comment (amenasria) [inline tasks/pipeline.py:827]: **💬 suggestion:** ‏A move invoke-way of doing this would be to use the `Exit` class with a color message (for a better visibility in the logs). Feel free to ignore if you want a `RuntimeError` instead !
```
from invoke.exceptions import Exit

from tasks.libs.common.color import Color, color_message
```
```suggestion
        raise Exit(color_message(f"Impossible to find the version of image {image_name} in circleci configuration file", Color.RED), code=1)
```

File: .github/workflows/buildimages-update.yml [hunk 11]
```diff
@@ -100,3 +104,43 @@ jobs:
           # the action fetches all branches and tags, in our case the branches we care about are already fetched
           # if they exist, so we can skip the fetch
           skip_fetch: true
+
+      - name: Check if PR exists
+        id: check_pr
+        env:
+          GH_TOKEN: ${{ github.token }}
```
Comment (amenasria) [inline .github/workflows/buildimages-update.yml:111]: **❓ question:** ‏Does this exist ? In the CI we're usually using `GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}`. I hope that this won't print the token as well

File: .github/workflows/buildimages-update.yml [hunk 12]
```diff
@@ -100,3 +104,43 @@ jobs:
...
+          if [ $prs -eq 0 ]; then
+            echo "CREATE_PR=true" >> $GITHUB_OUTPUT
+          fi 
+
+      # Create PR only if there is no pre-existing PR on the branch
+      - name: Create PR
+        if: ${{ steps.check_pr.outputs.CREATE_PR == 'true' }}
+        env:
+          TMP_PR_BODY_PATH: /tmp/pr_body
+          GH_TOKEN: ${{ github.token }}
```
Comment (amenasria) [inline .github/workflows/buildimages-update.yml:130]: **❓ question:** ‏Same question as above

**Review by amenasria — APPROVED** [review]
Left some comments, otherwise LGTM !

File: .github/workflows/buildimages-update.yml (hunk 11 above)
Comment (amenasria) [reply .github/workflows/buildimages-update.yml:111]: [Nevermind !](https://stackoverflow.com/a/66903734)

File: .github/workflows/buildimages-update.yml (hunk 12 above)
Comment (amenasria) [reply .github/workflows/buildimages-update.yml:130]: [Nevermind !](https://stackoverflow.com/a/66903734)

File: .github/workflows/buildimages-update.yml (hunk 1 above)
Comment (pgimalac) [reply .github/workflows/buildimages-update.yml:79]: I guess the issue is that `.gitlab-ci.yml` contains several entries which could be different...

File: .github/workflows/buildimages-update.yml (hunk 4 above)
Comment (pgimalac) [inline .github/workflows/buildimages-update.yml:146]: Could you use long options or add comments to make it clear what each is ?

File: tasks/pipeline.py [hunk 13]
```diff
@@ -810,6 +810,27 @@ def update_circleci_config(file_path, image_tag, test_version):
...
 
+@task(
+    help={
+        "file_path": "path of the CircleCI configuration YAML file",
+    },
+    autoprint=True,
+)
+def get_gitlab_config_image_tag(_, file_path=".gitlab-ci.yml"):
+    """
+    Print the current image tag of the given CircleCI configuration file (default: ".circleci/config.yml")
```
Comment (KevinFairise2) [inline tasks/pipeline.py:821]: Thanks for addressing my comment!
You probably need to update the comment and the help of the function to reflect the change

**Review by KevinFairise2 — APPROVED** [review]

File: tasks/buildimages.py [hunk 14]
```diff
@@ -33,3 +33,42 @@ def update_test_infra_definitions(ctx: Context, commit_sha: str, go_mod_only: bo
...
+
+### Buildimages update  
+This PR updates the current buildimages (`{old_build_image_tag}`) to `{new_build_image_tag}`{test_version_str}, [here is the full changelog]({compare_url}).  
+"""
+
+    if old_go_version != new_go_version:
+        pr_body += f"""
+
+### Golang update
+This PR updates the current Golang version ([`{old_go_version}`](https://go.dev/doc/devel/release#go{old_go_version})) to [`{new_go_version}`](https://go.dev/doc/devel/release#go{new_go_version}).
```
Comment (pgimalac) [inline tasks/buildimages.py:72]: Same as above, could you extract the URLs into separate variables to simplify the string ?

File: tasks/pipeline.py [hunk 15]
```diff
@@ -810,6 +810,27 @@ def update_circleci_config(file_path, image_tag, test_version):
         circle.write(circle_ci.replace(f"{match.group(0)}", f"{image}:{image_tag}\n"))
 
 
+@task(
+    help={
+        "file_path": "path of the CircleCI configuration YAML file",
```
Comment (pgimalac) [inline tasks/pipeline.py:815]:
```suggestion
        "file_path": "path of the Gitlab configuration YAML file",
```

**Review by pgimalac — APPROVED** [review]
